RFC: change lowering of ccall cconvert arguments and add Ref{T} type

vtjnash opened the pull request:
continuation of #9913, but targeted for merging into master

(note: this is part of the implementation of Julep https://gist.github.com/vtjnash/7296634)

vtjnash committed:
eliminate pointertoref Intrinsic and just use jl_value_ptr function reversably

vtjnash committed:
change lowering of ccall cconvert arguments

ccall(:a, B, (C,), d) is now lowered as:
  d#1 = cconvert_gcroot(C, d)
  Expr(:call, :ccall, :a, B, (C,), cconvert(C, d#1), d#1)

This allows implementation of the special-case ccall lowering logic in Julia
and still create a gcroot for it as needed

the Ref{T} type is the new supertype of Ptr{T}. the ccall emission
turns a Ref{T} into a Ptr{T}, as if the &addressOf operator had been
applied. the specific rules for this will be moved from the current
gist to the documentation at a later point.

JeffBezanson commented:
Overall this is really good. I think we can rename `cconvert_gcroot` to `cconvert`; the name is ugly and there aren't really any meaningful definitions of `cconvert` as it is, so we might as well merge them. There is also too much complexity in the several subtypes of `Ref`. They are somewhat redundant with `SubArray`, and I think it is easy enough to simply avoid anything like `pointer(allocate_array(), n)`. I like the PR better when it was a net deletion of code:)

vtjnash commented:
yeah, it may make more sense to just use SubArray instead. part of the intent of this PR is that you can replace most usages of `pointer` with `Ref`. essentially, `Ref` is intended to act as a gc-safe version of `Ptr` for almost anywhere

there is one meaningful cconvert call, which is `cconvert(Ptr{T}, Ref{T})`. there may be lots of valid `cconvert_gcroot` functions

JeffBezanson commented:
Yes, the fact that `cconvert_gcroot` is the more useful function to define is part of the problem: the good function has the ugly name.

I'm not sure it's worth trying to replace Ptr with Ref. A Ref is not a pointer, and is not as efficient as a pointer, so nobody will be fooled.

vtjnash commented:
> I'm not sure it's worth trying to replace Ptr with Ref. A Ref is not a pointer, and is not as efficient as a pointer, so nobody will be fooled

but a `Ref` is much more versatile and is often easy to allocate on the stack. i'm hoping to fool people someday

JeffBezanson commented on base/abstractarray.jl:
```diff
@@ -6,6 +6,17 @@ typealias AbstractVecOrMat{T} Union(AbstractVector{T}, AbstractMatrix{T})
 
+# convert Arrays to pointer arrays for ccall
+cconvert_gcroot{P<:Ptr}(::Type{Ptr{P}}, a::Array{P}) = a
+function cconvert_gcroot{P<:Ptr}(::Type{Ptr{P}}, a::Array)
```
Should go in array.jl as it is Array-specific.

vtjnash commented:
actually, there might be a few more useful calls for the method currently named `cconvert`, namely those with a first argument of `Ptr{T}` for some T. in general, it's not ideal to allow arrays and strings and such to be implicitly converted to `Ptr` on assignment. those operations should probably only be allowed in the context of ccall. however, it is fine to allow them to implicitly convert to a Ref object.

vtjnash commented:
> They are somewhat redundant with SubArray, and I think it is easy enough to simply avoid anything like pointer(allocate_array(), n)

i'm more concerned about whether it remains valid under all code transforms. for example, in the following somewhat-contrived example, the string has become unrooted by the time it reached function `i`, after the inlining pass has taken care of optimizing away the function call to `g(x)` and corresponding gc-root for `x`:
```julia
f() = g("hello world")
g(x) = i(h(x))
@noinline h(x) = pointer(x)
@noinline i(x) = ccall(:write, Cint, (Ptr{Uint8},), x)
```

JeffBezanson commented:
But if you can still call `pointer`, adding an extra layer of Ref doesn't
fix this. Granted, if nobody ever called `pointer` we'd be fine, but that's
true with or without Ref.

vtjnash commented:
true, but i would suggest most usages of `pointer` that exist now could be replaced by a `Ref` object construction instead. then the conversion to a `Ptr` would only implicitly happen during the `ccall` construction, when codegen is able to ensure, under all possible code transformations, that there has been a gc-root constructed (even though that gc-root may be actually on the stack and thus transiently destroyed very shortly thereafter via https://github.com/JuliaLang/julia/pull/8134)

vtjnash commented:
do you want me to merge this separately, or is it acceptable if I continue this under #7906 and interleave these two changes when it gets merge

vtjnash committed:
remove unused Ref subtypes. disallow implicit convert to Ptr{T} (ref #6618)

JeffBezanson commented on base/pcre.jl:
```diff
@@ -100,6 +101,7 @@ end
+    errstr[1] = C_NULL
     extra = ccall((:pcre_study, :libpcre), Ptr{Void},
                   (Ptr{Void}, Int32, Ptr{Ptr{UInt8}}),
                   regex, options, errstr)
```
Looks like bug fixes in this file? Could you apply these to master?

vtjnash replied:
It's not exactly a bug. Pcre should zero this field, but I had errors in the way I lowered ccall and this helped me get further

JeffBezanson commented on base/refpointer.jl:
```diff
@@ -0,0 +1,121 @@
+    if isbits(T)
+        return convert(P, data_pointer_from_objref(b))
+    else
+        return convert(P, data_pointer_from_objref(b.x))
```
I realize this does the correct thing for ccall, but it's a confusing semantic for a wrapper object to have: sometimes the wrapper pretends not to be there. I would prefer one of these:

1. Only allow constructing these for isbits types.
1. If `x` isn't a bits type just return it and don't wrap it.
1. Rename these types, and have `Ref` be just a simple assignable cell with the obvious behavior.

vtjnash replied:
I don't know how to make that behavior friendly to type inference. Is it possible, or perhaps even already functional?

vtjnash replied:
It's much better semantically that this is an abstract type, since it avoids a lot of ambiguity when specifying a type signature

vtjnash replied:
> I realize this does the correct thing for ccall, but it's a confusing semantic for a wrapper object to have: sometimes the wrapper pretends not to be there

also, this wrapper currently always exists and always pretends not to be there, such that it simply points to memory of a type specified by the type parameter when converted to a pointer.

also, if I don't wrap non-bits types, then i have to define `Base.convert{T}(P::Type{Ptr{T}}, b::T)` for all types, and I think that is a really bad idea.

note that there are some further refinements of this on the jn/ccall3 branch PR

vtjnash commented:
i think the is ready to merge once CI checks pass

vtjnash commented:
hm, i'm not really sure why travis linux is the only place this PR seems to fail

JeffBezanson commented:
Looks like one of those mysterious hangs. Heck, if it even works on *windows*, it must be right, right? :)

JeffBezanson commented:
* Yay for net deletion.
* I'm worried that removing so many conversions to `Ptr` is a pretty big breaking change.
* Needs documentation / comments

JeffBezanson commented on base/inference.jl:
```diff
@@ -133,7 +133,20 @@ t_func[fpiseq] = (2, 2, cmp_tfunc)
+            return Any
+        end
+        t = rt.parameters[1]
+        if isa(t,DataType) && is((t::DataType).name,Ref.name)
```
What does it mean if a ccall returns a `Ref`?

vtjnash replied:
See gist or doc updates on the jn/ccall3 PR. Some of these updates are primarily only meaningful in that context

JeffBezanson commented on base/base.jl:
```diff
@@ -52,16 +52,12 @@ convert{T}(::Type{(T...)}, x::Tuple) = cnvt_all(T, x...)
-cconvert{P<:Union(Int8,UInt8)}(::Type{Ptr{P}}, s::AbstractString) = bytestring(s)
+# conversions used by ccall
+ptr_arg_cconvert_gcroot{T}(::Type{Ptr{T}}, x) = cconvert_gcroot(T, x)
+ptr_arg_cconvert{T}(::Type{Ptr{T}}, x) = cconvert(T, x)
```
Would we eventually replace these with

```
convert{T,S}(::Type{Ref{T}}, r::Ref{S}) = Ref(convert(T, r[]))
```
?

vtjnash replied:
No. ptr_arg should eventually go away

JeffBezanson replied:
Will you be able to pass `Ref(1)` to a C function expecting a `int16_t*`, for example? I imagined that might be implemented by `Ref` conversions. This is really useful for calling fortran.

vtjnash replied:
Yes (Already implemented here)

Or pass 1 to an argument of type Ref{Int16}

vtjnash commented:
> Looks like one of those mysterious hangs. Heck, if it even works on windows, it must be right, right? :)

that sounds great, but it seemed to happen on every commit build, so i don't know what to make of it or how to repeat it

vtjnash committed:
add missing argument verification in some rand! calls

vtjnash commented:
now that I found the bug (affecting travis, but not source builds) and fixed it in https://github.com/JuliaLang/julia/commit/fdad916a4f4e95bf1e4d055560175efab71672fa, i should be able to merge this now, pending a green-light from travis

> I'm worried that removing so many conversions to Ptr is a pretty big breaking change.

perhaps, but that was sort of the point of deleting them

> Needs documentation / comments

it's written, and will merged as part of #7906

> bikeshedding @JeffBezanson 

do you want me to change the names of any of these methods (`Base.cconvert`, `Base.cconvert_gcroot`)?

vtjnash commented on src/jltypes.c:
```diff
@@ -3244,21 +3244,26 @@ void jl_init_types(void)
                                         jl_any_type, jl_null, 32);
 
     tv = jl_tuple1(tvar("T"));
+    jl_ref_type =
```
@JeffBezanson should we explicitly disable gc in this `jl_init_types` function? it would be bad if it somehow ran, even though it shouldn't have managed to allocate that much memory by here

JeffBezanson replied:
GC is already disabled at this point in the init process. I guess we could assert at the top of the function that GC is disabled.

vtjnash replied:
nah, just wanted to make sure i wasn't missing something

JeffBezanson commented:
I don't think we can just sneak in removing all these conversions to Ptr. cc @StefanKarpinski @ViralBShah 

Also, `convert` and `cconvert` is a much better set of names than `convert`, `cconvert`, and `cconvert_gcroot`. In fact that last name is so ugly that I veto it.

JeffBezanson commented:
Part of the problem with the name is that in reality, `cconvert_gcroot` is a relatively high-level function that converts a julia structure to a different one more suitable for consumption by C code. This behavior is the main purpose, and it's not directly related to the GC. Sure it's result will be rooted, but so is everything passed to a ccall.

vtjnash commented:
> I don't think we can just sneak in removing all these conversions to Ptr

Oh, I don't intend to be sneaky about it. It was quite a deliberate move :)

FWIW, at least one usage of explicit `convert` in base (that I deleted here) was dropping a gc-root. Arguably, it's much clearer to explicitly call `pointer` anyways (and much easier to find all usages in the code that way too).

The primary motivation is that I really don't like that the follow code works currently, while discarding two essential gc-roots along the way.
```julia
type Z
  x::Ptr{UInt32}
end

function f()
  local a::Ptr{UInt8}
  a = UInt8[1, 3, 4]
  return Z(UInt32[4,5,6])
end
```

I want this code pattern to to throw an error. I would put this in same category of `convert(UInt, typemax(Int))` also throwing an error. (seems that someone agrees with me too: https://github.com/JuliaLang/julia/pull/6618#issuecomment-41290444)

> Part of the problem with the name is that in reality, cconvert_gcroot is a relatively high-level function that converts a julia structure to a different one more suitable for consumption by C code. This behavior is the main purpose, and it's not directly related to the GC. Sure its result will be rooted, but so is everything passed to a ccall.

hm, that makes sense. i'm not sure if it's appropriate to call it `cconvert` either though, since it doesn't generally convert to the specified type. in that sense, `convert` and `cconvert` are close cousins right now (since both return an object of the type specified by the first argument). currently, `cconvert` is a weird function that calls convert, except for a few cases where it knows that ccall.cpp is going to manually and laboriously do the convert and then it passes through an object of the wrong type. there's lots of crazy branching that goes on in ccall to try to handle the cases where it has an object of the wrong type thrown at it, but which it was somehow supposed to handle anyways. 

perhaps my problem is that i don't fully know what `cconvert_gcroot` is as a more general concept, since it's really only meaningful as an intermediate step intended for creating a gcroot, in the context of a later call to `cconvert`.

some more examples, of code currently not written that just came to mind that could make use of this infrastructure includes:
* making it possible to pass a general substring to C, by making a copy exactly when needed:
```
function cconvert_gcroot{T}(::Type{Ptr{T}}, s::SubString)
  if endof(s) is endof(s.originalstring) and isbytestring(s.originalstring)
     return s
  else
     return bytestring(s)
end
```
* and something similar might be applicable for Stefan's utf8 work, so that we might write a method like: `cconvert_gcroot(::Type{Ptr{UInt8}}, s::UTF8String) = convert(::Array, s)`

note that currently, this only allows one additional level of gc-rooting. so you can't easily have `Array{String}` turn into `Array{Ptr}`

JeffBezanson commented:
That makes sense; I do like the idea of removing such dangerous implicit conversions.

I think there is a problem with the gc rooting here. A call like `ccall(f, R, (Ptr{Ptr{UInt8}},), ["a","b"])` will convert the new vector using `cconvert_gcroot`, but then only the converted version is rooted, and not the original. In the current system, the original is also rooted.

vtjnash commented:
alright, i think all that suggests that i should add one more builtin type:
```julia
immutable GCRootsC{T} <: Any
    c::T
    roots::Any
end
```
then i can eliminate `cconvert_gcroot` in favor of `cconvert(::Type{T}, x)` returning either `::T` or `::GCRootsC{T}`

will definitely need to think about this some more however. it's rather difficult to make this general, and ensure that enough gc-roots are always created, but still try to avoid creating extraneous ones too often.

vtjnash committed:
add an optional "roots" field to RefArray for tracking any additional required references

vtjnash commented:
@JeffBezanson bump. with the latest commit, converting an Array to a Ptr{Ptr} will maintain the gc-root for both the pointer array and the non-pointer (gc-roots) array

JeffBezanson commented on base/array.jl:
```diff
@@ -18,6 +18,29 @@ call{T}(::Type{Matrix{T}}, m::Integer, n::Integer) = Array{T}(m, n)
+        return RefArray(a,1,false) # root something, so that this function is type-stable
+    else
+        ptrs = Array(P, length(a)+1)
+        roots = Array(Any, length(a))
```
Ouch this is kind of expensive.

vtjnash replied:
in common cases, `cconvert_gcroot` is identity, but `cconvert` might be arbitrarily expensive. this just moves some of that work early and ensures `cconvert` is simple.

this is also significantly more general (and better rooted) than we what we have now. the old code could handle turning arrays of strings into pointers to pointers, but i don't entirely know what it would do given anything else.

JeffBezanson replied:
I don't think we were missing any roots, since before we allocated these objects in the temp_arg_area.

JeffBezanson commented:
Still need to get rid of `cconvert_gcroot`. One possibility is to have `cconvert` always make a `Ref` for Ptr arguments, then allow a `Ref` to be `convert`ed to Ptr. However I'm not even sure that's necessary. If ccall first calls cconvert, then convert, we can still implement conversions to Ptr in `cconvert` and avoid having them be part of `convert`.

vtjnash commented:
that doesn't work. it's missing the `_gcroot` part of the functions behavior

i could have both `cconvert` and `cconvert_gcroot` call `convert` as a fallback

however, the methods are now much simpler. `cconvert_gcroot` does one thing (create an object for gc-rooting so that it is safe to pass to `cconvert`) and `cconvert` does one thing (call `convert`, with added methods for converting to Ptr)

JeffBezanson commented:
No, I think it could work. `cconvert` could do whatever transformation is needed, e.g. calling `bytestring`, and also wrap the result in a Ref if necessary. Its result would be rooted. Then everything else can be handled by `convert`.

vtjnash commented:
I don't like it, since it means the Type{Ptr} methods need to be added back to the implicit "convert" method

JeffBezanson commented:
Yes, but only for `Ref`.

vtjnash commented:
No, all of them. Plus I need to add another new type that wraps a random object and turns it into a pointer when you call convert on it.

JeffBezanson commented:
But `cconvert` could make a `Ref` for anything that needed to be converted to `Ptr`, allowing conversions to `Ptr` to be defined only for `Ref`. This at least seems possible.

As it stands, this has ccall depend on *four* functions plus `convert`, and it creates garbage where we didn't previously. So I'm not sure this is enough of a net improvement.

vtjnash commented:
But you are proposing making more garbage Ref objects to avoid defining one functio no? (Also, I don't count the ptr_arg functions since this will also deprecate all that machinery)

JeffBezanson commented:
Not necessarily. For objects like `Vector{Int}` that don't need much translation, `cconvert` can still just return a Ptr, and have the following `convert` be a no-op. If more translation is needed, `cconvert` can return a `Ref` or other object that gets rooted, followed by the actual conversion to Ptr by `convert`.

vtjnash commented:
But what would convert(::String) return? Plus that means I need to add another variable to hold your new proposed gc root

JeffBezanson commented:
Converting a string could return a Ref to the Vector{UInt8} inside a ByteString.

Wouldn't this be the same variable currently used for `cconvert_gcroot`?

vtjnash commented:
that is true for ByteString, but it doesn't generalize particularly well. for example, what would you do for `PyObject`?

vtjnash committed:
renames cconvert(cconvert_gcroot()) to unsafe_convert(cconvert())

this also makes changes such that cconvert is mostly unchanged from its
functionality before this change. cconvert is expected to typically call
convert. in some cases (such as when intending to create a Ptr for
C-compatiblity), it should defer the actual unsafe step of making the
Ptr to unsafe_convert.

any code intending to allow conversion of an object to a Ptr should
define the operation as a method of unsafe_convert after this commit,
rather than the current behavior of making them a method of convert

vtjnash commented on base/refpointer.jl:
```diff
@@ -0,0 +1,77 @@
+Base.getindex(b::RefArray) = b.x[b.i]
+
+Base.setindex!(b::RefValue, x) = (b.x = x; b)
+Base.setindex!(b::RefArray, x) = (b.x[b.i] = x; b)
```
should i undefine the `setindex!` operation for a RefArray with roots? thoughts?

JeffBezanson replied:
Yes definitely. Arguably getindex is not needed either.

vtjnash replied:
well, there does need to be some way of turning a `Ref` back into a value

vtjnash replied:
i was thinking more along the lines of adding:
`Base.setindex!{T}(b::RefArray{T,Void}, x) = (b.x[b.i] = x; b)`
`Base.setindex!{T}(b::RefArray{T}, x) = error("this Ref has associated GC-roots and does not know how to do setindex! for them")`

JeffBezanson replied:
-1
I don't know why you'd need to setindex through a ref, and if you did that error message is pretty scary.

vtjnash replied:
to set an initial value, or populate an output of a function call, if used as an output argument

vtjnash commented:
this is mostly just waiting for CI to turn green to merge now

JeffBezanson commented:
Maybe we should remove the `pointer` function as well.

JeffBezanson commented:
This seems to make the system image significantly larger (more than 1MB, about 6%). Probably due to the lowering of ccall I'd guess.

vtjnash commented:
that's probably true. i removed the `pointertoref` intrinsic, but i'm thinking it might be necessary to reverse that and remove `jl_value_ptr`, to give a more concise AST after lowering

vtjnash commented:
> Maybe we should remove the pointer function as well.

i'm on board with that, although it would need to be deprecated for a time first

JeffBezanson commented:
I wouldn't have thought the increase was due to `jl_value_ptr`, just the extra unsafe_convert call. `pointertoref` wasn't used that much.

vtjnash commented:
perhaps not then. the lowering of `Ref` gets a bit verbose right now, for something that eventually turns into a single llvm bitcast. but it's true that base doesn't use it much right now
